README.txt
FlightGear MultiPlayer Server Cross Feed Client
-----------------------------------------------

File: README.txt - 20130124 - 20130123 - 20121024 - Started 20121017

A simple UPD client acting as a CROSSFEED for fgms. 

Author: Geoff R. McLane - <reports _AT_ geoffair _DOT_ info>
Licence: GNU GPL Version 2 (or later, at your choice)

Operation:
----------

Adjust the fgms.conf configuration file to add the address of 
this crossfeed - like -
crossfeed.host = 192.168.1.21
crossfeed.port = 3333

Then run fgms with this new configuration. It will immediately
dispatch a copy of EACH mutiplayer packet to this address 
by UDP.

When running cf_client add this configuraion to the command line
 --IP addr     (-I) = Set IP address to connect to fgms. (def=IPADDR_ANY)
 --PORT val    (-P) = Set PORT address to connect to fgms. (dep=3333)

like> cf_client -I 192.168.1.21 -P 3333

To see the full command line option run using -?, like -
C:\FG\17\build-cf>Release\cf_client.exe  -?
==========================================================================
cf_client.exe - version 1.0.25, compiled Jan 24 2013, at 19:06:23

fgms connection:
 --IP addr      (-I) = Set IP address to connect to fgms. (def=IPADDR_ANY)
 --PORT val     (-P) = Set PORT address to connect to fgms. (dep=3333)

Available IO:
 --ADDRESS ip   (-A) = Set IP address for Telnet and HTTP. (def=127.0.0.1)
 --TELNET port  (-T) = Set port for telnet. (def=0)
 --HTTP   port  (-H) = Set port for HTTP server (def=3335)
Note Telnet and HTTP will use the same address of 127.0.0.1.
Telnet or HTTP can be disabled using a 0 (or negative) port value.

File Outputs:
 --log file     (-l) = Set the log output file. (def=tempcft.txt)
 --json file    (-j) = Set the output file for JSON. (def=none)
 --raw file     (-r) = Set the packet output file. (def=none)
 --tracker file (-t) = Set the tracker output file. (def=none)
A file output can be disabled by using 'none' as the file name.
Relative file names will use the cwd [C:\FG\17\build-cf]

Tracker: Using PostgreSQL
 --DB name      (-D) = Set db name, 'none' to disable. (def=crossfeed)
 --ip addr      (-i) = Set db ip address. (def=127.0.0.1)
 --port num     (-p) = Set db port value. (def=5432)
 --user name    (-u) = Set db user name. (def=crossfeed)
 --word pwd     (-w) = Set db password. (def=crossfeed)

Miscellaneous:
 --help     (-h, -?) = This HELP and exit 0
 --air          (-a) = Set to modify AIRCRAFT string. (def=Off)
 --LIVE secs    (-L) = Set the flight TTL in integer secs. (def=10)
 -v[n]               = Bump or set verbosity - 0,1,2,5,9 (def=1)
 --conf file    (-c) = Load configuration from an INI type file.
See 'example.ini' in the source for valid sections and keys.
Prior to the command line a $HOME/.crossfeedrc INI file will be loaded,
if it exists. In Windows, that will be %USERPROFILE%\.crossfeedrc
==========================================================================


Ouput:
------

A list of currently active flights can be obtains by a simple 
GET request on the HTTP port, like -
 http://192.168.1.21:3335/flights.json - get the json string
 http://192.168.1.21:3335/flights.xml  - get the xml string

These strings are only updated each second, so GET requests more 
frequent than 1 Hz are a waste of resources. 


Building From Source:
---------------------

This crossfeed client (cf) uses CMake building.

The build.zip source file may contain some 'tools' to assist in this,
but they may need to be adjusted to your particular environment, and 
paths.

Create a new out-of-source build folder, say a sub-directory, 'build'
Change into this folder, and run -

For unix:

1: build $ cmake .. [options]
2: build $ make
3: build $ [sudo] make install (if desired)

For msvc:

Command Line Building:
1: Establish the MSVC environment, then
2: build-cf > cmake .. [Options, like -G 'generator']
3: build-cf > cmake --build . --config Release

GUI Building:
1: Load cmake-gui, and establish the source directory, and the binary 
diectory to thie new out-of-source build folder
2: Click [Configure], and select the 'generator'
3: Maybe adjust items and [Configure] again
4: Click [Generate]
5: Load cf_client.sln into MSVC, switch to Release configuration,
   and build the project (F7).

This process should create the cf_client binary.

Alternatively, in the 'build' folder, unzip the build.zip
build $ unzip ../build.zip
This file contains a number of tools, like build-fgx-cf.sh (and bat)
so for building it can be used -
build $ ./build-fgx-cf.sh [options]
Use -? to list the options this script supports.

And there is a script to clean all built components from this build
folder
build $ ./cmake-clean


List of CMake options: see CMakeLists.txt
-----------------------------------------

option USE_SIMGEAR_LIB Def=ON
# WARNING: Compile fixes would be needed to turn this OFF

option USE_POSTGRESQL_DATABASE Def=ON
option USE_SQLITE3_DATABASE    Def=OFF
# Tracker - at present only one or the other can be enabled

option ADD_HTTP_TEST Def=ON
# Only needed for test_http tool

option USE_GEOGRAPHIC_LIB Def=OFF
# Experimental - http://geographiclib.sourceforge.net/html/ 
# to test replacing SG

option BUILD_SERVER Def=OFF
# Presently not used, and NO SUPPORT for this option!

Source File List:
-----------------

README.txt         - this file
LICENCE.txt        - GNU GPL Version 2
CMakeLists.txt     - CMake 2.6 or later build

build.zip          - Some CMake build and other tools, scripts.
This should only be unzipped in an out-of-source build sub-directory. Each of 
the tools, scripts may need adjust to your particular environment.

src - <DIR> - main application
src\cf_version.hxx - Version file
src\cf_client.cxx  - main() - does all socket handling
src\cf_client.hxx
src\cf_pilot.cxx   - Deal with the packets from fgms crossfeed
src\cf_pilot.hxx
src\cf_ini.cxx     - Load configuration from an INI file
src\cf_ini.hxx
src\fip.h          - INI implemetation header

src\cf_pg.cxx      - If tracker data to postgresql database enabled
src\cf_pg.hxx

src\sqlite3.cxx    - If tracker data to sqlite3 database enabled.
src\sqlite3.hxx

Presently tracker data to a database can only be one or the other.

src\cf_server.cxx  - A 'test' server - NOT USED
src\cf_server.hxx
src\cf_discards.cxx - Not included in compile. Some discarded code

src\cf_lib - <DIR> - cf_lib  - miscellaneous function library (static)
src\cf_lib\cf_misc.cxx       - miscellanious function
src\cf_lib\cf_misc.hxx
src\cf_lib\mpKeyboard.cxx    - strobe the keyboard
src\cf_lib\mpKeyboard.hxx
src\cf_lib\mpMsgs.hxx        - the 'packet' from fgms
src\cf_lib\netSocket.cxx     - low level socket handling
src\cf_lib\netSocket.h
src\cf_lib\sprtf.cxx         - Output log
src\cf_lib\sprtf.hxx
src\cf_lib\tiny_xdr.hxx      - packet header
src\cf_lib\typcnvt.hxx       - convert to string function

Unused - Replaced with SimGearCore.lib
src\cf_lib\cf_euler.cxx      - euler_get() - Convert Orientation to heading, pitch, roll
src\cf_lib\cf_euler.hxx
src\cf_lib\fg_geometry.cxx   - class Point3D
src\cf_lib\fg_geometry.hxx
src\cf_lib\SGMath2.hxx       - some vectors used

PostgreSQL Tracker Enabled 
src\cf_lib\cf_postgres.cxx
src\cf_lib\cf_postgres.hxx

SQLite Tracker Enabled
src\cf_lib\cf_sqlite.cxx
src\cf_lib\cf_sqlite.hxx

Windows ONLY
src\cf_lib\win_strptime.cxx

Tools: src/tools - various testing tools
cf_feed.cxx  - read raw log and crossfeed to cf_client
test_http.cxx/hxx - just testing HTTP handling
test_pg.cxx - test posgresql connection and adding/deleting records
test_sqlite.cxx/hxx - test sqlite adding/enumerating/deleting records
test_pg2.cxx - test postgresql, and using geographic library
sqlLi2Pg.cxx/hxx - experimental transfer of records - not completed.


Dependencies:
-------------

Simple    - SimGearCore.lib, plus the usual 'system' libraries.
+Tacker   - PostgreSQL or SQlite3, and pthread library.
+test_pg2 - Requires Geographic library.


Test HTML Files:
----------------

src\test\index.html    - index to the following HTML
src\test\list.html     - JSON feed in simple table
src\test\map.html      - JSON feed displayed on OSM map
src\test\airports.html - Add an airport overlay to map
src\test\geodesic.htm  - Some examples of geodesic calculations
src\test\ol_map.html   - Simple example, drawing a track on a map.
plus the src\test subdirectories and files like -
css, images, img, and js

Enjoy.
Geoff.
20130124

# eof


ChangeLog.txt
2013/09/03 12:13:01 geoff

    * Expand the /flights.json URL to allow parameter following a '?'. These 
      parameters are entirely ignore. RFC by Pete.
    * Bump version to 1.0.29
    
2013/04/11 19:16:47 geoff

    * The inevitable minor fixes to compile in Ubuntu
    * Changed the unix build tools to use a sub-directory build/ which 
      is included in .gitignore, so git will ignroe all in this folder

2013/04/04 15:00:05 geoff

    * Add an XML pilot list feed on URL 'GET /flights.xml HTTP...'. The is 
      the SAME data as the json feed, except in XML format complinant with
      mpmap from http://mpmap01.flightgear.org/fg_server_xml.cgi?mpserver12.flightgear.org:5001
      however roll and pitch have been removed, and spd_kt="123" (TAS) added. Also 
      spd_kt, alt, heading are given as rounded up integers only. This is updated each second.
    * To conform to this URL style the json feed is on 'GET /flights.json HTTP...", but 
      also remains on /data.
    * Methods other than GET are now replied to with 405 Method Not Allowed, adding 
      "Allow : GET\r\n" to the reply if GET other than /flights.json, /flights.xml, /data, /info or /
    * The MAX_PACKET_SIZE set to 1200 (from 1024) to agree with FG multiplayermgr.cxx 
      which has had this value since before Sept 2008, the then CVS source.
    * A version file added containing the version string which is read in by the CMakeLists.txt
      as is now done in SG/FG. CMakeLists.txt will abort if NO version file found.
    * Enhanced test_pg, especially test 2, to output 'CLOSED' flights, and each related positions
      entry for each FID. Points out a considerable number of somewhat 'useless' records, where there
      are no position records, or just a few, or the aircraft did not really move. Writes a big 
      json file of the results of the queries, with timing...
    * Updated test_http to respond to the same URI as cf_client, passing back canned data for 
      testing.
    * Version bumped to 1.0.28


2013/02/05 12:09:38

    * Check the creation of a json file, if enabled, and abort if unable
      to create a blank file.
    * More fixes to ensure a file name is returned by get_..., else pointer to 
      'none'.
    * Small fix to ensure tracker log disabled by default.
    * In unix, add app exit time UTC if daemon and by signal.
    * Correct nanosleep converting seconds to milliseconds in windows port of
      this function.
    * Bumped AGE to 26.

2013/01/27 11:37:43

    * Change the cmake project name to 'crossfeed'. Seems this
      does vrey little in unix, and only changes the MSVC solution 
      file name in windows.
    * Changed CMake option USE_GEOGRAPHIC_LIB to OFF by default, removing
      dependency to Geographic library needed only for test_pg2
    * Tested removing SimgearCore library dependence, but several code 
      fixes required to build with this OFF.
    * Bumped AGE to 26.

2013/01/25 14:05:13

    * Remove pwd output from log file, and tidy up several 
      log messages during command line parsing.
    * For unix add #define stricmp strcasecmp
    * Minor addjustment in fip.h. Change if (line[0] == NULL) to
      if (line[0] == 0), to avoid gcc warning. There remains 
      a few gcc compiler warnings using -Wall, but do not seem 
      to effect running. Put a post on Google code, feather-ini-parser,
      to maybe even remove these.
      
2013/01/24 16:54:38

    * Implement new feature to read a configuration INI 
      file with command -c file.ini. Added new files -
      cf_ini.cxx/hxx. INI parsing is using FIP (feather-ini-parsing)
      from : https://code.google.com/p/feather-ini-parser/
      Code license GNU GPL v2, here included as fip.h
    * Added an example.ini to show the sections and key values
      available. In the main, it matches all that can be done 
      on the command line.
    * A number of other minor changes to fully accomodate the 
      use of this INI file - cf_client.cxx/hxx, cf_discards.cxx,
      cf_misc.cxx, sprtf.cxx/hxx
    * Prior to scanning the command line, seek and load 
      a $HOME/.crossfeedrc INI file to set defaults and 
      be able to run without any command line.
    * Updated README.txt to reflect current version.
    * Bumped AGE to 25.

2013/01/22 00:00:00

    * Main README.txt update
    * Correct name of db_setup README.sql to txt
    * Add some 'test' pages for crossfeed, etc
    * Bring crossfeed uptodate with fgx-cf
    
2013/01/05 14:27:55

    * Due to a ' character in the callsign broke the sql statement,
      restored some callsign modification. Now accept ALL alpha-numeric 
      chars, plus '-' and '_'.
    * Bumped AGE to 24.

2013/01/04 12:56:01

    * Allow '-D none' to DISABLE PostgreSQL tracker database.
    * Bumped AGE to 23.

2013/01/03 11:18:32

    * On database init, ensure all OPEN flights are CLOSED
    * Only install the primary app, cf_client
    * Bumped AGE to 22.

2012/12/30 15:26:01

    * Remove quotes from around json numbers
    * Add listen, http and telent addr/port to 'info' json
    * Add UI command to cf_feed.cxx
    * Bumped AGE to 21.

2012/12/29 13:17:11

    * Set default modifying callsign and aircraft to OFF.
    * Add cf_feed, to be able to feed raw packets to cf_client from a 
      raw packet log, without the need to run fgms.
    * Bumped AGE to 20.

2012/12/28 10:43:53

    * Due to temporary outage of fgx.ch server, establish a repo at 
      gitorious.org/fgtools to be able to devel in both linux and WIN32
    * Cut in http handling from test_http.cxx
    * Bumped AGE to 19.

2012/12/22 18:34:40

    * Add a URI of /info to get a json information string. At present 
      this contains minimal information, but can be easily extended 
      with all sorts of internal app information.
    * The bug using just a '/' persists, and a 404 is returned.
    * And the bug adding new fid for continuing flights persists, but 
      it has been reduced.

src/cf_ini.cxx
/*
 *  Crossfeed Client Project
 *
 *   Author: Geoff R. McLane <reports _at_ geoffair _dot_ info>
 *   License: GPL v2 (or later at your choice)
 *
 *   Revision 1.0.0  2013/01/23 00:00:00  geoff
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation; either version 2 of the
 *   License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful, but
 *   WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, US
 *
 */

// Module: cf_ini.cxx
#ifdef _MSC_VER
#pragma warning ( disable : 4996 ) // POSIX name depreciated
#endif
#include "cf_client.hxx"
#include "cf_pilot.hxx"
#include "cf_ini.hxx"
#include "fip.h"
#ifdef USE_POSTGRESQL_DATABASE
#include "cf_pg.hxx"
#else // !USE_POSTGRESQL_DATABASE
#ifdef USE_SQLITE3_DATABASE
#include "cf_sqlite3.hxx"
#endif
#endif // USE_POSTGRESQL_DATABASE y/n
#include "sprtf.hxx"
#include "cf_misc.hxx"
#include "cf_ini.hxx"

#ifndef WIN32
#define stricmp strcasecmp
#endif

/* --------------------------------------
   <example.ini>
##########################################################################
# Example INI file for cf_client
# cf_client - from version 1.0.25, Jan 24 2013
# line that begin with hash, '#', are comments
# Loaded by cf_client by -c example.ini
# The order of this command on the command line will determine if these 
# value override any previous commands, but later commands on the 
# command line will prevail.
# This content can be placed in a $HOME/.crossfeedrc file to run 
# without a command line... it is loaded first.
# Note: 'bool' values should use 0 for false/off, 1 for true/on
# section names, keys and values can NOT contain spaces
# Spaces after the key and after the '=' before the value are allowed.
##########################################################################

# fgms connection:
[connection]
# --IP addr      (-I) = Set IP address to connect to fgms. (def=IPADDR_ANY)
address=ANY
# --PORT val     (-P) = Set PORT address to connect to fgms. (dep=3333)
port=3333

# Available IO:
[IO]
# --ADDRESS ip   (-A) = Set IP address for Telnet and HTTP. (def=127.0.0.1)
address=127.0.0.1
# --TELNET port  (-T) = Set port for telnet. (def=0)
telnet=0
# --HTTP   port  (-H) = Set port for HTTP server (def=3335)
http=3335
# Note Telnet and HTTP will use the same address of 127.0.0.1.
# Telnet or HTTP can be disabled using a 0 (or negative) port value.

# File Outputs:
[outputs]
# --log file     (-l) = Set the log output file. (def=tempcft.txt)
log_file=tempcft.txt
# --json file    (-j) = Set the output file for JSON. (def=none)
json_file=none
# --raw file     (-r) = Set the packet output file. (def=none)
raw_file=none
# --tracker file (-t) = Set the tracker output file. (def=none)
tracker_file=none
# A file output can be disable by using 'none' as the file name.

# Tracker: Using PostgreSQL
[postgresql]
# --DB name      (-D) = Set db name, 'none' to disable. (def=crossfeed)
database=crossfeed
# --ip addr      (-i) = Set db ip address. (def=127.0.0.1)
address=127.0.0.1
# --port num     (-p) = Set db port value. (def=5432)
port=5432
# --user name    (-u) = Set db user name. (def=crossfeed)
user=crossfeed
# --word pwd     (-w) = Set db password. (def=crossfeed)
password=crossfeed

# Tracker: Using SQLite3
[sqlite]
# --sql name     (-s) = Set SQLite3 db file, 'none' to disable.
database=none

# NOTE: At this time only ONE or the OTHER of these 'tracker'
# databases can be active.

[miscellaneous]
# --air          (-a) = Set to modify AIRCRAFT string. (def=Off) USE 0=false/off, 1=true/on
mod_aircraft=1
# --LIVE secs    (-L) = Set the flight TTL in integer secs. (def=10)
live_secs=10
# -v[n]               = Bump or set verbosity - 0,1,2,5,9 (def=1)
verbosity=1
# --daemon       (-d) = Run as daemon. (def=Off) 
# does NOT apply to Windows - USE 0=false/off, 1=true/on
daemon=1

# eof
##########################################################################
   -------------------------------------- */

int readINI( char *file )
{
    int iret = 0;
    std::string s;
    int i;
    //bool b;
    INI<> ini(file,false);
    if (!ini.Parse())
        return 1;   // FAILED to load INI file
    if (ini.Select("connection")) {
        // # --IP addr      (-I) = Set IP address to connect to fgms. (def=IPADDR_ANY)
        s = ini.Get<std::string,std::string>("address",get_crossfeed_address());
        if (stricmp(s.c_str(),"ANY"))
            set_crossfeed_address(s);
        else
            set_crossfeed_address("");
        // # --PORT val     (-P) = Set PORT address to connect to fgms. (dep=3333)
        i = ini.Get<std::string,int>("port", get_crossfeed_port());
        set_crossfeed_port(i);
    }
    //# Available IO:
    if (ini.Select("IO")) {
        //# --ADDRESS ip   (-A) = Set IP address for Telnet and HTTP. (def=127.0.0.1)
        s = ini.Get<std::string,std::string>("address",get_server_address());
        set_server_address(s); //address=127.0.0.1
        // # --TELNET port  (-T) = Set port for telnet. (def=0)
        i = ini.Get<std::string,int>("telnet", get_telnet_port());
        set_telnet_port(i); // telnet=0
        // # --HTTP   port  (-H) = Set port for HTTP server (def=3335)
        i = ini.Get<std::string,int>("http", get_http_port());
        set_http_port(i); // http=3335
        // # Note Telnet and HTTP will use the same address of 127.0.0.1.
        // # Telnet or HTTP can be disabled using a 0 (or negative) port value.
    }

    // # File Outputs:
    if (ini.Select("outputs")) {
        // # --log file     (-l) = Set the log output file. (def=tempcft.txt)
        s = ini.Get<std::string,std::string>("log_file",get_cf_log_file());
        set_cf_log_file((char *)s.c_str()); // if 'none' will disable log // log_file=tempcft.txt or cf_client.log

        // # --json file    (-j) = Set the output file for JSON. (def=none)
        s = ini.Get<std::string,std::string>("json_file",(json_file ? json_file : "none"));
        if (strcmp(s.c_str(),"none")) {
            json_file = strdup(s.c_str());
            json_file_disabled = false; //json_file=flights.json
        } else
            json_file_disabled = true; //json_file=none

        // # --raw file     (-r) = Set the packet output file. (def=none)
        s = ini.Get<std::string,std::string>("raw_file",get_raw_log());
        if (strcmp(s.c_str(),"none")) {
            set_raw_log( s.c_str() ); // { raw_log = strdup(log); } // "cf_raw.log";
            set_raw_log_disable( false ); // { raw_log_disabled = set; } // raw_file=cf_raw.log
        } else
            set_raw_log_disable( true ); // { raw_log_disabled = set; } // raw_file=none

        // # --tracker file (-t) = Set the tracker output file. (def=none)
        s = ini.Get<std::string,std::string>("tracker_file",get_tracker_log());
        if (strcmp(s.c_str(),"none")) {
            set_tracker_log( s.c_str() ); // { tracker_log = strdup(log); } // "cf_tracker.log";
            set_tracker_log_disable( false );
        } else
            set_tracker_log_disable( true ); // { tracker_log_disabled = set; } // tracker_file=none
        // # A file output can be disable by using 'none' as the file name.
    }
#ifdef USE_POSTGRESQL_DATABASE
    if (ini.Select("postgresql")) {
        //printf("\nTracker: Using PostgreSQL\n");
        //printf(" --DB name      (-D) = Set db name, 'none' to disable. (def=%s)\n", get_pg_db_name() );
        s = ini.Get<std::string,std::string>("database",get_pg_db_name());
        if (strcmp(s.c_str(),"none")) {
            set_pg_db_name((char *)s.c_str());
            Enable_SQL_Tracker = true; // tracker database enabled
        } else
            Enable_SQL_Tracker = false; // tracker database disabled
        //printf(" --ip addr      (-i) = Set db ip address. (def=%s)\n", get_pg_db_ip() );
        s = ini.Get<std::string,std::string>("address",get_pg_db_ip());
        set_pg_db_ip((char *)s.c_str());
        //printf(" --port num     (-p) = Set db port value. (def=%s)\n", get_pg_db_port() );
        s = ini.Get<std::string,std::string>("port",get_pg_db_port());
        set_pg_db_port((char *)s.c_str());
        //printf(" --user name    (-u) = Set db user name. (def=%s)\n", get_pg_db_user() );
        s = ini.Get<std::string,std::string>("user",get_pg_db_user());
        set_pg_db_user((char *)s.c_str());
        //printf(" --word pwd     (-w) = Set db password. (def=%s)\n", get_pg_db_pwd() );
        s = ini.Get<std::string,std::string>("password",get_pg_db_pwd());
        set_pg_db_pwd((char *)s.c_str());
    }
#endif // #ifdef USE_POSTGRESQL_DATABASE y/n
#ifdef USE_SQLITE3_DATABASE // ie !#ifdef USE_POSTGRESQL_DATABASE
    if (ini.Select("sqlite")) {
        //printf("\nTracker:\n");
        //printf(" --sql name     (-s) = Set SQLite3 db file, 'none' to disable. (def=%s)\n", get_sqlite3_db_name() );
        s = ini.Get<std::string,std::string>("database",get_sqlite3_db_name());
        if (strcmp(s.c_str(),"none")) {
            set_sqlite3_db_name((char *)s.c_str());
            Enable_SQL_Tracker = true;
        } else
            Enable_SQL_Tracker = false; // tracker database disabled
    }
#endif // #ifdef USE_SQLITE3_DATABASE // ie !#ifdef USE_POSTGRESQL_DATABASE

    if (ini.Select("miscellaneous")) {
        // # --air          (-a) = Set to modify AIRCRAFT. (def=Off)
        i = ini.Get<std::string,int>("mod_aircraft",(int)m_Modify_AIRCRAFT);
        m_Modify_AIRCRAFT = i ? true : false; // b; // mod_aircraft=off
        // # --LIVE secs    (-L) = Set the flight TTL in integer secs. (def=10)
        i = ini.Get<std::string,int>("live_secs", (int)m_PlayerExpires);
        m_PlayerExpires = i; // live_secs=10
        //# -v[n]               = Bump or set verbosity - 0,1,2,5,9 (def=1)
        i = ini.Get<std::string,int>("verbosity", verbosity);
        verbosity = i;  //verbosity=1
        // # --daemon       (-d) = Run as daemon. (def=Off)
        i = ini.Get<std::string,int>("daemon",RunAsDaemon);
        RunAsDaemon = i;
    }

    return iret;
}

int loadRCFile()
{
    char *home = getenv("HOME");
    if (!home)
        home = getenv("USERPROFILE"); // XP=C:\Documents and Settings\<name>, Win7=C:\Users\<user>
    if (home) {
        std::string rc(home);
#ifdef _MSC_VER
        rc += "\\";
#else
        rc += "/";
#endif
        rc += CF_RC_FILE;
        if (is_file_or_directory((char *)rc.c_str()) == DT_FILE)
            return readINI((char *)rc.c_str());
    }
    return 0;
}


// eof - cf_ini.cxx
